ultravox/training/train.py

```diff
@@ -317,12 +317,17 @@ def train(args: config_base.TrainConfig):
         logging.info(f"train end time: {t_end}")
         logging.info(f"elapsed: {t_end - t_start}")
 
-    if is_master:
-        # Saving the model using pipeline to ensure its code is saved
-        pipeline = ultravox_pipeline.UltravoxPipeline(
-            model, tokenizer=text_tokenizer, device=device
-        )
-        pipeline.save_pretrained(args.output_dir)
+    # Save the pipeline code and update the config to include the pipeline
+    pipeline = ultravox_pipeline.UltravoxPipeline(
+        model, tokenizer=text_tokenizer, device=model.device
+    )
+    # We don't want to save the model twice. Trainer.save_model saves the model to the output_dir.
```

Comment:
But why is a standalone `save_model` call needed at all?

Reply:
Because `save_model` knows how to save FSDP models correctly but pipeline does not (see attached pieces of code).
You could argue that the same might be replicable with pipeline given the right context object and that would be fair. I finally got something working and didn't want to touch it 😅

I can take another pass at it.

<img width="792" alt="image" src="https://github.com/user-attachments/assets/62cfc0e7-edb0-4e89-b414-efcd00ccbac0">
<img width="636" alt="image" src="https://github.com/user-attachments/assets/c0f10cea-f6b3-4720-9603-24132dc3f192">

Reply:
I did some more hacking around and I still don't see a way out of using the `trainer.save_model` call.

The `pipeline` however, is only needed for saving the pipeline code and updating the config.
I could copy and paste the following code, which would look nicer in the short-term, but I feel like it's safer not to do that (in case `transformers`' way of doing things changes):
![image](https://github.com/user-attachments/assets/415e0956-adc4-4c33-9a86-06497e405323)

Reply:
OK, I suggest just clarifying the comment to explain this, ie that the pipeline doesn't know how to save FSDP models properly, so we need to do it manually.